library/core/src/main/java/com/google/android/exoplayer2/text/cea/Cea708Decoder.java

```diff
@@ -270,81 +300,113 @@ private void finalizeCurrentPacket() {
+        if (serviceNumber < 7) {
+          Log.w(TAG, "Invalid extended service number: " + serviceNumber);
+        }
+        currOffset++;
       }
-    }
 
-    // Ignore packets in which blockSize is 0
-    if (blockSize == 0) {
-      if (serviceNumber != 0) {
-        Log.w(TAG, "serviceNumber is non-zero (" + serviceNumber + ") when blockSize is 0");
+      // Ignore packets in which blockSize is 0
+      if (blockSize == 0) {
+        if (serviceNumber != 0) {
+          Log.w(TAG, "serviceNumber is non-zero (" + serviceNumber + ") when blockSize is 0");
+        }
+        return;
       }
-      return;
-    }
 
-    if (serviceNumber != selectedServiceNumber) {
-      return;
-    }
-
-    // The cues should be updated if we receive a C0 ETX command, any C1 command, or if after
-    // processing the service block any text has been added to the buffer. See CEA-708-B Section
-    // 8.10.4 for more details.
-    boolean cuesNeedUpdate = false;
-
-    while (serviceBlockPacket.bitsLeft() > 0) {
-      int command = serviceBlockPacket.readBits(8);
-      if (command != COMMAND_EXT1) {
-        if (command <= GROUP_C0_END) {
-          handleC0Command(command);
-          // If the C0 command was an ETX command, the cues are updated in handleC0Command.
-        } else if (command <= GROUP_G0_END) {
-          handleG0Character(command);
-          cuesNeedUpdate = true;
-        } else if (command <= GROUP_C1_END) {
-          handleC1Command(command);
-          cuesNeedUpdate = true;
-        } else if (command <= GROUP_G1_END) {
-          handleG1Character(command);
-          cuesNeedUpdate = true;
-        } else {
-          Log.w(TAG, "Invalid base command: " + command);
+      if (serviceNumber == selectedServiceNumber) {
+        int b = 0;
+        while ( b < blockSize) {
+          // start accumulating the service blocks
+          if (currOffset >= currentDtvCcPacket.currentIndex) {
+            // extra protection to avoid array out of bound.
+            break;
+          }
+          serviceBlockList.add(currentDtvCcPacket.packetData[currOffset++]);
+          b++;
         }
-      } else {
-        // Read the extended command
-        command = serviceBlockPacket.readBits(8);
-        if (command <= GROUP_C2_END) {
-          handleC2Command(command);
-        } else if (command <= GROUP_G2_END) {
-          handleG2Character(command);
-          cuesNeedUpdate = true;
-        } else if (command <= GROUP_C3_END) {
-          handleC3Command(command);
-        } else if (command <= GROUP_G3_END) {
-          handleG3Character(command);
-          cuesNeedUpdate = true;
+        // Convert the ArrayList to byte[].
+        byte[] blk = new  byte[serviceBlockList.size()];
+        for (int i = 0; i < serviceBlockList.size(); i++) {
+          blk[i] = serviceBlockList.get(i);
+        }
+        serviceBlockPacket.reset(blk, serviceBlockList.size());
```

Comment:
I might not be following exactly what's happening here (forgive my relative inexperience with CEA-708), but it seems like each time you go round the outer while loop (that starts on L314), if you find a service block with the selected number you add its bytes to serviceBlockList and then allow that to be parsed/consumed by the (mostly existing) inner while loop that starts on L363.

But if you find multiple (N) service blocks with a matching service number (maybe that's impossible...?) then you'll parse the first one N times, the second N-1 etc.

Something like:
```
selectedServiceNum=2
currentDtvCcData = {
  blockA { num=1 },
  blockB { num=2 },
  blockC { num=3 },
  blockD { num=2 },
}
```

Will cause the while loop on L363 to first be executed with:
```
serviceBlockPacket = {
  blockB { num=2 },
}
```

And then again with:
```
serviceBlockPacket = {
  blockB { num=2 },
  blockD { num=2 },
}
```

a) Have I understood what the code is doing correctly
b) Is that intended/necessary?

I have a copy of the CEA-708-B spec (dated Dec 1999) - so if there's sections in there that you can point to to help me understand what's meant to be going on that might be helpful!

Reply:
Thank you @icbaker for looking at this. I think I can split 1 and 3 into a separate PR. Though 3 aligns more with 4 and 5 but coding wise it looks independent. But for 2, it depends on restructuring of serviceBlockPacket and I think its better if we handle 2, 4 and 5 together. I am open for suggestions.  

For the scenario -
```
selectedServiceNum=2
currentDtvCcData = {
  blockA { num=1 },
  blockB { num=2 },
  blockC { num=3 },
  blockD { num=2 },
}
```
Will cause the while loop on L363 to first be executed with:
```
serviceBlockPacket = {
  blockB { num=2 },
}
```
Before we proceed to next iteration of outer loop, the L407 clears the  serviceBlockList.
```
if (ccDataState == CC_DATA_STATE_COMPLETE_PARAM) {
  serviceBlockList.clear();
}
```
And then, the blockD is added.  Then the inner while loop is called with -  
```
serviceBlockPacket = {
  blockD { num=2 },
}
```
With that, blockB and blockD are decoded as expected.
 
The outer loop adds all the service blocks from the selected service number to serviceBlockList. The purpose of serviceBlockList is to handle the commands with multiple parameters. For example DefineWindow command (CEA-708-D  section 8.10.5.2 DEFINE WINDOW - (DF0 … DF7)) has 6 parameters each of 1 byte. It makes Define Window coding is of total 7 bytes. All these 7 bytes can either be part of single service block or can be spread across multiple service blocks.  ParsableBitArray does not provide a way to add/extended an existing object.  ParsableBitArray.reset() is going to clear the previous data and leads to disassociation of command and its associated parameters. 

In this fix, if the serviceBlock ends before it provides all the parameters required for the command, the ccDataState is set to CC_DATA_STATE_WAITING_PARAM and the inner while loop breaks at LN376 and the serviceBlockList is not cleared at LN408. In the next iteration of the outer while loop, the next service block is appended to serviceBlockList. The serviceBlockList now contains 2 service blocks which together describes a command and its parameters. The inner while loop now has all the information to decode a command.

For example
```
selectedServiceNum=2
currentDtvCcData = {
  blockA { num=1 },
  blockB { num=2,  has partial DF command  like “0x9A, 0x38, 0x4A, 0xD1” },
  blockC { num=3 },
  blockD { num=2 has the next parameters for command 0x9A which are like “0x8B, 0x0F, 0x11”},
}
```
Will cause the while loop on L363 to first executed with:
```
serviceBlockPacket = {
  blockB { num=2  with “0x9A, 0x38, 0x4A, 0xD1”},
}
```
The condition at LN 590 becomes true
```
if (serviceBlockPacket.bitsLeft() < DFS_COMMAND_LEN) {
  ccDataState = CC_DATA_STATE_WAITING_PARAM;
  break;
}
```
That makes cuesNeedUpdate to remain false and skips serviceBlockList.clear() at LN408. The serviceBlockList now holds blockB.
Then the blockD (without the service block header) is appended to serviceBlockList on the next iteration of outer while loop. Now the inner loop is called with 
```
serviceBlockPacket = {
 blockB { num=2 }
 blockD { num = 2 }
}
```
This now describes command DF (0x9A) with all 6 parameters and ccDataState becomes CC_DATA_STATE_COMPLETE_PARAM. This time the new cue is added.

I am referring to  CEA-708-D from August 2008. The https://en.wikipedia.org/wiki/CEA-708 also provides a good overview.

Reply:
@icbaker points 1 and 3 are separated into PR #8415.